### PR TITLE
[ROCm][DSA] Enable AITER fused FP8 indexer writer

### PR DESCRIPTION
**Pull request by fanxingran**
## Motivation

On HIP, the DSA FP8 indexer writer currently issues a chain of small kernels (K LayerNorm, joint Q/K RoPE, optional Hadamard, Q/K FP8 quant, and the paged index-K cache store). That launch overhead shows up on decode ITL, especially the tail.

CUDA already fuses the analogous indexer writer (`SGLANG_DISABLE_DSA_INDEXER_FUSION`, default on). This PR adds the ROCm counterpart via AITER `indexer_qk_rope_quant_and_cache`, so HIP can fuse RoPE + K LayerNorm + FP8 quant + cache write into one kernel. Hadamard is omitted, matching the CUDA fused writer (pre-quant logits stay aligned).

The path is **default-on** when AITER is available and indexer geometry matches (`head_dim=128`, `rope_dim=64`, `quant_block=128`, index-K LayerNorm). Models that do not match (including RMSNorm index-K) keep the legacy split writer. Set `SGLANG_AITER_FUSED_FP8_DSA_INDEXER=0` to force the split path.

Validated on GLM-5.2-MXFP4 (gfx950). Any DSA indexer with the same geometry can take this path; it is not hardcoded to a model name.

## Modifications

- `python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py`: AITER adapter (geometry checks, RoPE table reshape, fused Q/K write).
- `python/sglang/srt/layers/attention/dsa/dsa_indexer.py`: enable fused writer when HIP + AITER + geometry match; LayerNorm in fp32 only on the fused path; warn only if the env is explicitly set and unsupported.
- `python/sglang/srt/environ.py`: `SGLANG_AITER_FUSED_FP8_DSA_INDEXER = EnvBool(True)`.
- Tests:
  - `test/registered/unit/layers/attention/test_fp8_fused_writer_hip.py`
  - `test/registered/amd/test_fp8_fused_dsa_writer_gfx950.py`

## Accuracy Tests

GSM8K via local `sglang.test.run_eval` (1314 questions), GLM-5.2-MXFP4, baseline = fused off (`SGLANG_AITER_FUSED_FP8_DSA_INDEXER=0`).

| | Baseline | Fused |
|---|---:|---:|
| Score | 0.935 | 0.936 |

No accuracy regression. Server log contains `Enabled AITER fused FP8 DSA indexer writer`.

## Speed Tests and Profiling

Hardware: gfx950. Workload: InferenceX AgentX, **1200s** (not the 3600s publication window). Toggle only `SGLANG_AITER_FUSED_FP8_DSA_INDEXER`. Interactivity = output token throughput per user; TPOT = ITL.

### TP4 / EP4 / concurrency 8, HiCache DRAM

Request mix is aligned (510 vs 518 completed; ISL/OSL mean 103925/568 vs 104042/563).

| Metric | Baseline | Fused | Delta |
|---|---:|---:|---:|
| P90 interactivity (tok/s/user) | 190.02 | 194.48 | **+2.4%** |
| P90 E2E-norm interactivity | 153.12 | 157.97 | **+3.2%** |
| P50 interactivity | 135.78 | 137.08 | +1.0% |
| TTFT p90 | 1.197 s | 1.225 s | +2.3% |
| TPOT p90 | 13.0 ms | 12.5 ms | **−3.8%** |
| TPOT avg | 9.19 ms | 8.64 ms | **−6.0%** |
| TPOT p99 | 40.5 ms | 34.1 ms | **−15.7%** |
| E2E latency p90 | 10.85 s | 9.80 s | −9.6% |
| Throughput / chip | 10832 tok/s | 11013 tok/s | +1.7% |
| Output throughput | 235.5 tok/s | 236.9 tok/s | +0.6% |
| Completed requests | 510 | 518 | — |

Gain is mainly decode tail latency (p99 ITL), not prefill.

### TP8 / EP1 / concurrency 1, no HiCache

Request count 155 vs 157; ISL aligned. **OSL mean 646 vs 582**, so E2E and output throughput are **not** causal. Use interactivity / ITL.

| Metric | Baseline | Fused | Delta |
|---|---:|---:|---:|
| P90 interactivity (tok/s/user) | 244.64 | 252.77 | **+3.3%** |
| P90 E2E-norm interactivity | 214.62 | 224.71 | **+4.7%** |
| P50 interactivity | 234.51 | 241.99 | +3.2% |
| TTFT p90 | 1.689 s | 1.557 s | **−7.8%** |
| TPOT p90 | 4.64 ms | 4.53 ms | **−2.3%** |
| TPOT p99 | 7.20 ms | 6.92 ms | **−4.0%** |
| Completed requests | 155 | 157 | — |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (env is documented in `environ.py`; no cookbook change.)
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #35710656934](https://github.com/sgl-project/sglang/actions/runs/35710656934)
Latest PR Test (Extra): :x: [Run #35710656598](https://github.com/sgl-project/sglang/actions/runs/35710656598)
Latest PR Test (AMD ROCm 10): :hourglass_flowing_sand: [Run #35710656849](https://github.com/sgl-project/sglang/actions/runs/35710656849)

### CODE REVIEW

**Comment by fanxingran** [comment]
Hi @kkHuang-amd @HaiShaw @yuan-luo — could you take a look at this ROCm DSA fused FP8 indexer writer (AITER `indexer_qk_rope_quant_and_cache`, default-on when geometry matches)?

CI failed at `pr-gate` only (~5s); the actual test jobs did not run. Could you please `/tag-and-rerun-ci` (and `/tag-and-rerun-ci extra` if needed) so the AMD pipeline can start?

Thanks!

**Comment by 1am9trash** [comment]
Hi, @fanxingran 
I added the tag and re-triggered CI, but I found that lint format check failed.
May you take a look when you are free? Thanks.

**Comment by fanxingran** [comment]
> Hi, @fanxingran I added the tag and re-triggered CI, but I found that lint format check failed. May you take a look when you are free? Thanks.

Hi @1am9trash, thanks for tagging CI and catching the lint.

I ran pre-commit locally (ruff format). It only reformatted wrapping in `dsa_indexer.py` and `test_fp8_fused_dsa_writer_gfx950.py` — no logic change. Pushed in `54362bf`. Local hooks are green; CI should pick this up.

Thanks!

**Comment by fanxingran** [comment]
Hi @1am9trash — we pushed a gfx950 arch gate (`a24a22de8d`) so MI300 no longer takes the fused writer (the AITER Python symbol can import there, but the GPU kernel is gfx950-only). Also dropped the HIP unit test from NVIDIA `base-a` CPU.

Our own push is rate-limited: `User 'fanxingran' already triggered 'PR Test Base' via 'pull_request' ... wait 120 minutes`. Could you please `/tag-and-rerun-ci` so Base/AMD pick up this commit? Thanks!

File: python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def _is_gfx95() -> bool:
+    # AITER's fused indexer writer is gfx950-only. gfx942 (MI300) may still
+    # import the Python symbol and then fault in the kernel.
+    if not getattr(torch.version, "hip", None):
+        return False
+    if not torch.cuda.is_available():
+        return False
+    return "gfx95" in str(torch.cuda.get_device_properties(0).gcnArchName)
```
Comment (1am9trash) [inline python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py]: Hi, @fanxingran 
Can `_is_gfx95()` reuse `sglang.srt.utils.is_gfx95_supported()` utility?

File: python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        )
+    if norm_weight.dtype != torch.float32 or norm_bias.dtype != torch.float32:
+        raise TypeError(
+            "fused FP8 writer reads LayerNorm parameters as fp32; got "
+            f"{norm_weight.dtype} and {norm_bias.dtype}"
+        )
+
+    from aiter.ops.cache import indexer_qk_rope_quant_and_cache
+
+    indexer_qk_rope_quant_and_cache(
```
Comment (1am9trash) [inline python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py:99]: Hi, @fanxingran 
This path calls `indexer_qk_rope_quant_and_cache`, which sglang indexer did not use before. 
Do we need to bump the sglang aiter version?

File: python/sglang/srt/environ.py [hunk 3]
```diff
@@ -1658,6 +1658,9 @@ class Envs:
     SGLANG_ENABLE_PCG_DSV2_DUAL_STREAM = EnvBool(False)
     SGLANG_DSA_TOPK_BROADCAST = EnvBool(False)
     SGLANG_DISABLE_DSA_INDEXER_FUSION = EnvBool(False)
+    # HIP analog of CUDA DSA indexer fusion (default on). Set 0 for the
+    # legacy split writer. Omits Hadamard; pre-quant logits match.
+    SGLANG_AITER_FUSED_FP8_DSA_INDEXER = EnvBool(True)
```
Comment (1am9trash) [inline python/sglang/srt/environ.py]: Hi, @fanxingran 
CUDA uses `SGLANG_DISABLE_DSA_INDEXER_FUSION` (default on). Should we use the same disable-style flag?

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py
Comment (1am9trash) [inline python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: Hi, @fanxingran 
On some fallbacks (`_get_q_k_bf16`), Hadamard is gated only by `_maybe_rotate()`, which does not include the HIP fused flag. 
May that crash our code path, or are those paths unreachable on HIP?

File: python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py (hunk 1 above)
Comment (fanxingran) [reply python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py]: Reused sglang.srt.utils.is_gfx95_supported() and dropped the local helper.

File: python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py (hunk 2 above)
Comment (fanxingran) [reply python/sglang/kernels/ops/attention/dsa/fp8_fused_writer_hip.py:99]: No sglang AITER pin in this PR. Images clone ROCm/aiter; if indexer_qk_rope_quant_and_cache is missing we fall back via ImportError/AttributeError.

File: python/sglang/srt/environ.py (hunk 3 above)
Comment (fanxingran) [reply python/sglang/srt/environ.py]: Added SGLANG_DISABLE_AITER_FUSED_FP8_DSA_INDEXER (default False). Kept it separate from SGLANG_DISABLE_DSA_INDEXER_FUSION because that CUDA flag also packs wk/weights_proj.

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py
Comment (fanxingran) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: _maybe_rotate now skips Hadamard when the HIP fused writer is on as well, so fallback K matches the fused cache. It would mismatch, not crash.

**Comment by fanxingran** [comment]
Hi @1am9trash, thanks for the review.

All four are addressed in `b4356e3`:
1. Reused `is_gfx95_supported()`.
2. No sglang AITER pin; missing `indexer_qk_rope_quant_and_cache` falls back via ImportError.
3. HIP uses `SGLANG_DISABLE_AITER_FUSED_FP8_DSA_INDEXER` (default on). Separate from CUDA `SGLANG_DISABLE_DSA_INDEXER_FUSION` because that flag also packs `wk`/`weights_proj`.
4. `_maybe_rotate` now skips Hadamard when the HIP fused writer is on.

Could you please take another look at those replies? If anything still needs a change, I will follow up right away; if it looks good, an approve would be very helpful.

CI on `b4356e3` is already running (Lint / Base / AMD).

**Review by 1am9trash — APPROVED** [review]
Thanks for the update.
All comments are resolved.
LGTM.

**Comment by fanxingran** [comment]
Hi @1am9trash, the AMD mi35x failure is in test_triton_mla_kv_cache.py, which is unrelated to this PR — we did not change Triton MLA / KV cache. That test should be fine; this looks like a flake. Could you please /rerun-failed-ci? Thanks!

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 4]
```diff
@@ -8,6 +8,12 @@
 from einops import rearrange
 
 from sglang.kernels.fused_op import BaseFusedOp
+from sglang.kernels.ops.attention.dsa.fp8_fused_writer_hip import (
+    aiter_fused_fp8_qk_write,
+    aiter_fused_fp8_writer_available,
+    fused_fp8_writer_geometry_supported,
+    prepare_aiter_rope_caches,
+)
```
Comment (HaiShaw) [inline python/sglang/srt/layers/attention/dsa/dsa_indexer.py:16]: @fanxingran let's put this hip specific import under _is_hip, in follow-up PR

**Review by HaiShaw — APPROVED** [review]

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py (hunk 4 above)
Comment (fanxingran) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py:16]: Thanks @HaiShaw，I will move the HIP-only import under _is_hip  in a follow-up.

**Comment by HaiShaw** [comment]
AITER specific

**Comment by HaiShaw** [comment]
Previous `base-c-test-8-gpu-b300 (0)` was known issue listed in https://github.com/sgl-project/sglang/issues/17050
